src/Interpreters/Cache/Metadata.cpp

```diff
@@ -249,91 +233,202 @@ LockedKeyPtr CacheMetadata::lockKeyMetadata(
+                continue;
+            }
+        }
+        ++it;
+    }
+}
+
+void CacheMetadata::removeKey(const Key & key, bool if_exists, bool if_releasable)
+{
+    auto metadata_lock = lockMetadata();
 
-    FileCacheKey cleanup_key;
-    while (cleanup_queue->tryPop(cleanup_key))
+    auto it = find(key);
+    if (it == end())
     {
-        auto it = find(cleanup_key);
-        if (it == end())
-            continue;
+        if (if_exists)
+            return;
+        else
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "No such key: {}", key);
+    }
+
+    auto locked_key = it->second->lockNoStateCheck();
+    if (locked_key->getKeyState() != KeyMetadata::KeyState::ACTIVE)
+    {
+        if (if_exists)
+            return;
+        else
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "No such key: {}", key);
+    }
+
+    bool removed_all = locked_key->removeAllFileSegments(if_releasable);
+    if (removed_all)
+        removeEmptyKey(it, *locked_key, metadata_lock);
+}
+
+CacheMetadata::iterator CacheMetadata::removeEmptyKey(iterator it, LockedKey & locked_key, const CacheMetadataGuard::Lock &)
+{
+    const auto & key = locked_key.getKey();
 
-        LockedKeyPtr locked_metadata;
+    if (!it->second->empty())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot remove non-empty key: {}", key);
+
+    locked_key.markAsRemoved();
+    auto next_it = erase(it);
+
+    LOG_DEBUG(log, "Key {} is removed from metadata", key);
+
+    const fs::path key_directory = getPathForKey(key);
+    const fs::path key_prefix_directory = key_directory.parent_path();
+
+    try
+    {
+        if (fs::exists(key_directory))
+            fs::remove_all(key_directory);
+    }
+    catch (...)
+    {
+        LOG_ERROR(log, "Error while removing key {}: {}", key, getCurrentExceptionMessage(true));
+        chassert(false);
+        return next_it;
+    }
+
+    try
+    {
+        std::unique_lock mutex(key_prefix_directory_mutex);
+        if (fs::exists(key_prefix_directory) && fs::is_empty(key_prefix_directory))
+            fs::remove(key_prefix_directory);
+    }
+    catch (...)
+    {
+        LOG_ERROR(log, "Error while removing key {}: {}", key, getCurrentExceptionMessage(true));
+        chassert(false);
+    }
+    return next_it;
+}
+
+class CleanupQueue
+{
+    friend struct CacheMetadata;
+public:
+    void add(const FileCacheKey & key)
+    {
+        bool inserted;
         {
-            ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::FilesystemCacheLockKeyMicroseconds);
-            locked_metadata = std::make_unique<LockedKey>(it->second);
+            std::lock_guard lock(mutex);
+            if (cancelled)
+                return;
+            inserted = keys.insert(key).second;
         }
-
-        const auto key_state = locked_metadata->getKeyState();
-        if (key_state == KeyMetadata::KeyState::ACTIVE)
+        chassert(inserted);
```

Comment:
Why do we have this assert? Are there any guarantees that we won't add the same key twice in the queue?

Reply:
yes

Reply:
Please add clarification comment. Because it's confusing to have an assert and right after if statement with the same condition. When I wrote the comment regarding checking if there is such key in the `keys` set, I didn't thought that we have such guarantees

Reply:
Added a comment.